### PR TITLE
list-ops: add test template (#1958)

### LINKED ISSUE #1958: list-ops: add test template
## Task
Create a test template file for `list-ops`

### Details
1. If this is your first time contributing to this repository, welcome!
    - Contributing to a repository on GitHub requires a [fork](https://help.github.com/en/articles/fork-a-repo)
1. Create a branch on your local repository called `list-ops-add-test-template`
    - Run: `git checkout -b list-ops-add-test-template`
1. Clone problem-specifications repository; this is used to populate the template
    - Run: `git clone https://github.com/exercism/problem-specifications ../problem-specifications`
1. Create `exercises/list-ops/.meta/template.j2`
    - See [Generator Documentation] for more information
1. Generate test file from new template
    - Run: `bin/generate_tests.py list-ops`

At this point, if the generator fails, or you are otherwise stuck,
please feel free to reach out! It is perfectly acceptable to submit
an incomplete PR in order to get feedback or assistance.

### Resources
- [Contributing Guide]
- [Generator Documentation]
- [Closing issues using keywords]

[Contributing Guide]: https://github.com/exercism/python/blob/master/docs/CONTRIBUTING.md
[Generator Documentation]: https://github.com/exercism/python/blob/master/docs/GENERATOR.md
[Closing issues using keywords]: https://help.github.com/en/articles/closing-issues-using-keywords

### PR DESCRIPTION
**Pull request by dvermd**
Resolves #1958

### CODE REVIEW

**Comment by cmccandless** [comment]
Please check in the template.j2 you created as well as any changes to the test file.

**Comment by dvermd** [comment]
> Please check in the template.j2 you created as well as any changes to the test file.

Done. I shouldn't have pushed that late.

**Comment by dvermd** [comment]
I don't know why the CI check failed. I ran it on my side and it finished with return code 0

File: exercises/list-ops/.meta/additional_tests.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
+  "cases": [
+    {
+      "description": "fold additional cases",
+      "cases": [
```
Comment (cmccandless) [inline exercises/list-ops/.meta/additional_tests.json]: Any reason we can't just single-nest these here?

**Comment by cmccandless** [comment]
> I don't know why the CI check failed. I ran it on my side and it finished with return code 0

The CI check failed because the generated results of your template do not match the changes you checked in. This is most likely to the following workflow:
- Create/edit template
- Generate tests
- Check in changes
- Make minor edit to template and check in the change
- [the test file should be generated and checked in again here]
- Push

File: exercises/list-ops/.meta/additional_tests.json (hunk 1 above)
Comment (dvermd) [reply exercises/list-ops/.meta/additional_tests.json]: There is no reason not to single nest these tests definitions. I just kept to the format of canonical-data. 
Should I unfold them ?

**Comment by dvermd** [comment]
> > I don't know why the CI check failed. I ran it on my side and it finished with return code 0
> 
> The CI check failed because the generated results of your template do not match the changes you checked in. This is most likely to the following workflow:
> 
>     * Create/edit template
> 
>     * Generate tests
> 
>     * Check in changes
> 
>     * Make minor edit to template and check in the change
> 
>     * [the test file should be generated and checked in again here]
> 
>     * Push

I understand the check process and it's a good thing. The problem I'm facing is that I deleted list_ops_test.py, regenerated it with ./bin/generate_tests.py -vv list-ops and git tells me nothing is to be staged. Thus I don't know what to do.
Here is my procedure

>  david@david-pc ~/dev/exercism/python $ git reset --hard HEAD^
> HEAD est maintenant à 8cf7db3 list-ops: add test template (#1958)
>  david@david-pc ~/dev/exercism/python $ md5sum exercises/list-ops/list_ops_test.py
> 7118e0b2d43ce28cc5c618c8d39d8c97  exercises/list-ops/list_ops_test.py
>  david@david-pc ~/dev/exercism/python $ git reset --hard origin/list-ops-add-test-template
> HEAD est maintenant à cf370b4 list-ops: add missing generator (#1958)
>  david@david-pc ~/dev/exercism/python $ rm exercises/list-ops/list_ops_test.py
>  david@david-pc ~/dev/exercism/python $ git status exercises/list-ops
> Sur la branche list-ops-add-test-template
> Modifications qui ne seront pas validées :
>   (utilisez "git add/rm <fichier>..." pour mettre à jour ce qui sera validé)
>   (utilisez "git restore <fichier>..." pour annuler les modifications dans le répertoire de travail)
> 	supprimé :        exercises/list-ops/list_ops_test.py
> aucune modification n'a été ajoutée à la validation (utilisez "git add" ou "git commit -a")
>  david@david-pc ~/dev/exercism/python $ bin/generate_tests.py -v list-ops
> INFO:generator:list-ops: attempting render
> DEBUG:generator:list-ops: formatting tmp file
> list-ops generated at exercises/list-ops/list_ops_test.py
>  david@david-pc ~/dev/exercism/python $ git status exercises/list-ops
> Sur la branche list-ops-add-test-template
> rien à valider, la copie de travail est propre
>  david@david-pc ~/dev/exercism/python $ md5sum exercises/list-ops/list_ops_test.py
> 7118e0b2d43ce28cc5c618c8d39d8c97  exercises/list-ops/list_ops_test.py`

**Comment by cmccandless** [comment]
Hmm... Can you please apply the following patch, check it in, and see if Travis passes?

```
diff --git a/bin/generate_tests.py b/bin/generate_tests.py
index 3737cd3..f30bdad 100755
--- a/bin/generate_tests.py
+++ b/bin/generate_tests.py
@@ -185,7 +185,7 @@ def generate_exercise(env, spec_path, exercise, check=False):
 
         if check:
             try:
-                if not filecmp.cmp(tmp.name, tests_path):
+                if not filecmp.cmp(tmp.name, tests_path, shallow=False):
                     logger.error(
                         f"{slug}: check failed; tests must be regenerated with bin/generate_tests.py"
                     )
```

File: exercises/list-ops/.meta/additional_tests.json (hunk 1 above)
Comment (dvermd) [reply exercises/list-ops/.meta/additional_tests.json]: I unfolded them...

**Comment by dvermd** [comment]
I changed the import order to make sure the template.j2 and the test.py were to be staged. I pushed them but still no luck with the CI. Any more hints ?

**Comment by cmccandless** [comment]
I'm not seeing anything to indicate why the CI `generate_tests.py --check` is failing. I cannot reproduce on my local machine.

@yawpitch, if you have the time to look closely at this, do you see anything out of order?

**Comment by yawpitch** [comment]
I think we may need to add some logging to determine, _before_ the “if check” block if both `tmp.name` and `tests_path` (ie the _test.py) file actually exist on disk, and if they exist what their `os.stat` values are.

Something being wrong there is the only way I see that check failing so silently and oddly, implying that _somehow_ we’ve gotten into a situation where either the temp file is disappearing or the test file hasn’t been cloned down.

**Comment by cmccandless** [comment]
Another idea I've been toying with is using [`difflib`](https://docs.python.org/3/library/difflib.html) instead of `filecmp` so that the output actually shows the difference between the generated and actual content.

**Comment by yawpitch** [comment]
> Another idea I've been toying with is using [`difflib`](https://docs.python.org/3/library/difflib.html) instead of `filecmp` so that the output actually shows the difference between the generated and actual content.

Possibility of vanishing files notwithstanding it probably makes the most sense to only do difflib when filecmp comparison fails, so maybe not “instead of”.

**Comment by cmccandless** [comment]
OK, @dvermd, I see the problem now.

You need to pull the latest for problem-specifications. The list-ops exercise was updated there since you cloned the repo.

**Comment by cmccandless** [comment]
Merged! @dvermd thanks for working on this!

**Pull request merged**